# Train.csv 

In [2]:
import pandas as pd
import datetime as dt

In [15]:
df_train = pd.read_csv("../wns-wizard-av/Train/train.csv")

df_train.head()

impression_id      impression_time  user_id  app_code  \
0  c4ca4238a0b923820dcc509a6f75849b  2018-11-15 00:00:00    87862       422   
1  45c48cce2e2d7fbdea1afc51c7c6ad26  2018-11-15 00:01:00    63410       467   
2  70efdf2ec9b086079795c442636b55fb  2018-11-15 00:02:00    71748       259   
3  8e296a067a37563370ded05f5a3bf3ec  2018-11-15 00:02:00    69209       244   
4  182be0c5cdcd5072bb1864cdee4d3d6e  2018-11-15 00:02:00    62873       473   

     os_version  is_4G  is_click  
0           old      0         0  
1        latest      1         1  
2  intermediate      1         0  
3        latest      1         0  
4        latest      0         0

## Understanding Data

In [16]:
df_train.shape

(237609, 7)

In [17]:
df_train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 237609 entries, 0 to 237608
Data columns (total 7 columns):
impression_id      237609 non-null object
impression_time    237609 non-null object
user_id            237609 non-null int64
app_code           237609 non-null int64
os_version         237609 non-null object
is_4G              237609 non-null int64
is_click           237609 non-null int64
dtypes: int64(4), object(3)
memory usage: 12.7+ MB


In [18]:
# Convert impression_time become date format
df_train['impression_time'] = pd.to_datetime(df_train["impression_time"], format = "%Y-%m-%d %H:%M:%S")
df_train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 237609 entries, 0 to 237608
Data columns (total 7 columns):
impression_id      237609 non-null object
impression_time    237609 non-null datetime64[ns]
user_id            237609 non-null int64
app_code           237609 non-null int64
os_version         237609 non-null object
is_4G              237609 non-null int64
is_click           237609 non-null int64
dtypes: datetime64[ns](1), int64(4), object(2)
memory usage: 12.7+ MB


### Information
* There is no error in assigning data type to dataset train.csv

In [21]:
df_train['impression_id'].describe()

count                               237609
unique                              237609
top       d67456b240769b4c16d711215220de5d
freq                                     1
Name: impression_id, dtype: object

In [22]:
df_train['impression_time'].describe()

count                  237609
unique                  36461
top       2018-12-04 00:01:00
freq                       46
first     2018-11-15 00:00:00
last      2018-12-13 07:43:00
Name: impression_time, dtype: object

In [30]:
# Check out userId at the time where the most userId is exposed
df_train[df_train['impression_time'] == pd.to_datetime("2018-12-04 00:01:00", format = "%Y-%m-%d %H:%M:%S")]['user_id'].value_counts()

22910    1
48465    1
67423    1
28832    1
29538    1
12698    1
69527    1
64342    1
81557    1
70356    1
64210    1
28048    1
85351    1
78543    1
63693    1
8394     1
70706    1
78944    1
27783    1
66374    1
28612    1
53505    1
42859    1
69804    1
7778     1
27748    1
6464     1
46714    1
43759    1
60794    1
22839    1
12086    1
35829    1
12404    1
43306    1
26802    1
40047    1
25838    1
79428    1
15468    1
875      1
71978    1
5289     1
71464    1
21863    1
73728    1
Name: user_id, dtype: int64

### Information 
* Time with the most user expose to the ad is **2018-12-04 00:01:00**
* There is no error in entry point about impression_time for there is no user with same userId at one time
* There is 36461 impression_time that has only one user exposed to the ad

In [84]:
print(df_train['impression_time'].dt.date.describe())
df_train[(df_train['impression_time'].dt.date == dt.datetime(2018,11,27).date()) & (df_train['is_click'] == 1)].describe()

count         237609
unique            29
top       2018-11-27
freq           14243
Name: impression_time, dtype: object


user_id    app_code       is_4G  is_click
count    591.000000  591.000000  591.000000     591.0
mean   45616.475465  241.991540    0.357022       1.0
std    26920.946453  129.645533    0.479527       0.0
min       98.000000    3.000000    0.000000       1.0
25%    21756.500000  145.000000    0.000000       1.0
50%    45781.000000  231.000000    0.000000       1.0
75%    70058.000000  371.000000    1.000000       1.0
max    92565.000000  512.000000    1.000000       1.0

### Information 
* Date with the most user expose to the ad is **2018-11-27** 14243
* And from 14243 user, there are 591 user that clicked the ad.

In [37]:
df_train[df_train['is_click'] == 1]['impression_time'].describe()

count                   10862
unique                   9069
top       2018-11-27 00:15:00
freq                        5
first     2018-11-15 00:01:00
last      2018-12-13 07:43:00
Name: impression_time, dtype: object

### Information
* There is 10862 user that clicked when exposed to the ad
* There are 5 users that clicked the ad at 2018-11-27 00:15:00, the largest number at this time


In [41]:
df_train[df_train['is_click'] == 1]['impression_time'].dt.date.describe()

count          10862
unique            29
top       2018-12-03
freq             600
Name: impression_time, dtype: object

In [31]:
df_train['user_id'].describe()

count    237609.000000
mean      46454.526828
std       26802.726666
min           0.000000
25%       23197.000000
50%       46597.000000
75%       69684.000000
max       92586.000000
Name: user_id, dtype: float64

In [56]:
pd.to_datetime("2018-11-27", format="%Y-%m-%d")

Timestamp('2018-11-27 00:00:00')

In [5]:
df_v = pd.read_csv("../wns-wizard-av/Train/view_log.csv")
df_v

server_time device_type  session_id  user_id  item_id
0        2018-10-15 08:58:00     android      112333     4557    32970
1        2018-10-15 08:58:00     android      503590    74788     7640
2        2018-10-15 08:58:00     android      573960    23628   128855
3        2018-10-15 08:58:00     android      121691     2430    12774
4        2018-10-15 08:58:00     android      218564    19227    28296
5        2018-10-15 08:58:00     android      635314    87387    86660
6        2018-10-15 08:58:00     android      482459      348    35370
7        2018-10-15 08:58:00     android      218564    19227   100065
8        2018-10-15 08:58:00     android      947548    30183    86558
9        2018-10-15 08:58:00     android      941129    31754   127900
10       2018-10-15 08:58:00     android       58445    76152    95823
11       2018-10-15 08:58:00     android      264366    34712    11871
12       2018-10-15 08:58:00     android      585539    39617    22859
13       2018-10-15 08:58:00     android      876311    35913    15434
14       2018-10-15 08:59:00     android      845216    16793     2715
15       2018-10-15 08:59:00     android      845216    16793     2715
16       2018-10-15 08:59:00     android     1075632     5859    17418
17       2018-10-15 08:59:00     android      104628    66973    99981
18       2018-10-15 08:59:00     android      573960    23628    33633
19       2018-10-15 08:59:00     android      843031    34503    18299
20       2018-10-15 08:59:00     android      503590    74788   129764
21       2018-10-15 08:59:00     android      600876    59049    51049
22       2018-10-15 08:59:00     android      712208    27170   126076
23       2018-10-15 08:59:00     android      230591    81687    98681
24       2018-10-15 08:59:00     android      260261    90761   132849
25       2018-10-15 08:59:00     android     1068088    90962     6641
26       2018-10-15 08:59:00     android      712208    27170    52488
27       2018-10-15 08:59:00     android       19064    40741     4086
28       2018-10-15 08:59:00     android      634927    24208    61679
29       2018-10-15 08:59:00     android     1005272    44420   126451
...                      ...         ...         ...      ...      ...
3118592  2018-12-11 23:59:00     android      180402    33424   107282
3118593  2018-12-11 23:59:00     android      135534    49480    83589
3118594  2018-12-11 23:59:00     android       12845    57286    48334
3118595  2018-12-11 23:59:00     android      991622    69686   116661
3118596  2018-12-11 23:59:00     android      748797    79064    22411
3118597  2018-12-11 23:59:00     android      896679     3692    83751
3118598  2018-12-11 23:59:00     android      304187    26419   109061
3118599  2018-12-11 23:59:00     android     1037504    81422    53411
3118600  2018-12-11 23:59:00     android      435077    11934    43978
3118601  2018-12-11 23:59:00     android      600428    48699    71700
3118602  2018-12-11 23:59:00     android     1001989    49985    95249
3118603  2018-12-11 23:59:00     android      774398    61923    68713
3118604  2018-12-11 23:59:00     android      522865    19140    57499
3118605  2018-12-11 23:59:00     android        2055     4727   119516
3118606  2018-12-11 23:59:00     android      482741    36109    32822
3118607  2018-12-11 23:59:00     android      784554    24396    81146
3118608  2018-12-11 23:59:00     android      386770    26430    30349
3118609  2018-12-11 23:59:00     android      937574    91313   130415
3118610  2018-12-11 23:59:00     android      482741    36109       70
3118611  2018-12-11 23:59:00     android      415068    38431    37894
3118612  2018-12-11 23:59:00     android      748797    79064    22411
3118613  2018-12-11 23:59:00     android      919844    52394    35625
3118614  2018-12-11 23:59:00     android      401151    36344    44686
3118615  2018-12-11 23:59:00     android      954211    43333   130046
3118616  2018-12-11 23:59:00     android 